cpp/mg_utility/mage.hpp

```diff
@@ -0,0 +1,2757 @@
+  bool operator!=(const Id &other) const { return !(*this == other); }
+
+  bool operator<(const Id &other) const { return id_ < other.id_; }
+
+ private:
+  explicit Id(int64_t id) : id_(id) {}
+
+  int64_t id_;
+};
+
+/// @brief Wrapper class for @ref mgp_graph.
+class Graph {
+ private:
+  friend class Node;
+  friend class Relationship;
+
+ public:
+  explicit Graph(mgp_graph *graph) : graph_(graph) {}
+
+  /// @brief Returns the graph order (number of nodes).
+  int64_t order() const;
+  /// @brief Returns the graph size (number of relationships).
+  int64_t size() const;
+
+  /// @brief Returns an iterable structure of the graph’s nodes.
+  GraphNodes nodes() const;
+  /// @brief Returns an iterable structure of the graph’s relationships.
+  GraphRelationships relationships() const;
+
+  /// @brief Returns the graph node with the given ID.
+  Node GetNodeById(const Id node_id) const;
+
+  /// @brief Returns whether the graph contains a node with the given ID.
+  bool Contains(const Id node_id) const;
+  /// @brief Returns whether the graph contains the given node.
+  bool Contains(const Node &node) const;
+  /// @brief Returns whether the graph contains the given relationship.
+  bool Contains(const Relationship &relationship) const;
+
+  /// @brief Returns whether the graph is mutable.
+  bool IsMutable() const { return mgp::graph_is_mutable(graph_); }
+  /// @brief Creates a node and adds it to the graph.
+  void CreateNode() { mgp::graph_create_vertex(graph_, memory); }
+  /// @brief Deletes a node from the graph.
+  void DeleteNode(const Node &node);
+  /// @brief Deletes a node and all its incident edges from the graph.
+  void DetachDeleteNode(const Node &node);
+  /// @brief Creates a relationship of type `type` between nodes `from` and `to` and adds it to the graph.
+  void CreateRelationship(const Node &from, const Node &to, const std::string_view type);
+  /// @brief Deletes a relationship from the graph.
+  void DeleteRelationship(const Relationship &relationship);
+
+ private:
+  mgp_graph *graph_;
+};
+
+/// @brief View of graph nodes; wrapper class for @ref mgp_vertices_iterator.
+class Nodes {
+ public:
+  explicit Nodes(mgp_vertices_iterator *nodes_iterator) : nodes_iterator_(nodes_iterator) {}
+
+  class Iterator {
+   public:
+    friend class Nodes;
+
+    explicit Iterator(mgp_vertices_iterator *nodes_iterator);
+    ~Iterator();
+    Iterator &operator++();
+    Iterator operator++(int);
+    bool operator==(Iterator other) const;
+    bool operator!=(Iterator other) const { return !(*this == other); }
+    Node operator*();
+    // Traits
+    using difference_type = Node;
+    using value_type = Node;
+    using pointer = const Node *;
+    using reference = const Node &;
+    using iterator_category = std::forward_iterator_tag;
+
+   private:
+    mgp_vertices_iterator *nodes_iterator_ = nullptr;
+    size_t index_ = 0;
+  };
+
+  Iterator begin();
+  Iterator end();
```

Comment:
```suggestion
  Iterator begin() const;
  Iterator end() const;
```

Reply:
You might also want to consider implementing const style iterators as well. I would say this article explains implementation iterators well https://www.internalpointers.com/post/writing-custom-iterators-modern-cpp

Reply:
Thanks! I’ve actually studied this article while working on this API :slightly_frowning_face: